### PR TITLE
Client: Working signature verification

### PR DESCRIPTION
**Pull request by gdbelvin**
This adds signature verification to the client.

It also separates `HashLogRoot` from `crypto/signer.go` since it is not generic. 
It also adds a generic `SignObject`, `VerifyObject` functions.

### CODE REVIEW

File: crypto/key_manager.go [hunk 1]
```diff
@@ -82,140 +59,84 @@ func (k PEMKeyManager) HashAlgorithm() crypto.Hash {
...
+
+	switch key.(type) {
+	case *ecdsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_ECDSA
+	case *rsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_RSA
+	default:
+		return nil, errors.New("unsupported key type")
```
Comment (codingllama) [inline crypto/key_manager.go]: nit: Add the key type to the error message?

**Comment by RJPercival** [comment]
Once you rebase this on master, we can take a look.

File: crypto/data_formats.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"encoding/base64"
+	"strconv"
+
+	"github.com/benlaurie/objecthash/go/objecthash"
+	"github.com/google/trillian"
+	"github.com/google/trillian/crypto/sigpb"
+)
+
+// This file contains struct specific mappings and data structures.
+// TODO: remove data-structure specific operations.
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	return hash[:]
+}
+
+// SignLogRoot updates a log root to include a signature from the crypto signer this object
+// was created with. Signatures use objecthash on a fixed JSON format of the root.
+func (s Signer) SignLogRoot(root trillian.SignedLogRoot) (*sigpb.DigitallySigned, error) {
+	objectHash := HashTrillianSignedLogRoot(root)
+	signature, err := s.Sign(objectHash[:])
+
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+// Constants used as map keys when building input for ObjectHash. They must not be changed
+// as this will change the output of hashRoot()
+const (
+	mapKeyRootHash       string = "RootHash"
+	mapKeyTimestampNanos string = "TimestampNanos"
+	mapKeyTreeSize       string = "TreeSize"
+)
+
+// HashTrillianSignedLogRoot hashes SignedLogRoot objects in a custom way.
+func HashTrillianSignedLogRoot(root trillian.SignedLogRoot) []byte {
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

**Comment by gdbelvin** [comment]
Rebased on master and fixed comments.  PTAL.

File: crypto/data_formats.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	// using JSON to verify hashes can build the exact same input to ObjectHash.
+	rootMap[mapKeyRootHash] = base64.StdEncoding.EncodeToString(root.RootHash)
+	rootMap[mapKeyTimestampNanos] = strconv.FormatInt(root.TimestampNanos, 10)
+	rootMap[mapKeyTreeSize] = strconv.FormatInt(root.TreeSize, 10)
+
+	hash := objecthash.ObjectHash(rootMap)
+	return hash[:]
+}
+
+// SignLogRoot updates a log root to include a signature from the crypto signer this object
```
Comment (RJPercival) [inline crypto/data_formats.go]: "SignLogRoot updates a log root" - I don't think it does. It just returns the signature, which is better than modifying the root IMO. Comment needs correcting to reflect what the func actually does.

File: crypto/data_formats.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+// Constants used as map keys when building input for ObjectHash. They must not be changed
+// as this will change the output of hashRoot()
+const (
+	mapKeyRootHash       string = "RootHash"
+	mapKeyTimestampNanos string = "TimestampNanos"
+	mapKeyTreeSize       string = "TreeSize"
+)
+
+// HashTrillianSignedLogRoot hashes SignedLogRoot objects in a custom way.
```
Comment (RJPercival) [inline crypto/data_formats.go]: "in a custom way" - could the comment elaborate on that a little please?

File: crypto/data_formats_test.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/trillian"
+)
+
+const (
+	publicKeyFile      = "../testdata/log-rpc-server.pubkey.pem"
+	privateKeyFile     = "../testdata/log-rpc-server.privkey.pem"
```
Comment (RJPercival) [inline crypto/data_formats_test.go]: These paths might not work inside of Google. Instead, could you import "github.com/google/trillian/testonly" and use testonly.Demo{Private,Public}Key?

File: crypto/data_formats_test.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+				TreeSize:       2,
+			},
+		},
+	} {
+		before := test.root // Capture test value.
+		signature, err := signer.SignLogRoot(test.root)
+		if err != nil {
+			t.Errorf("Failed to sign log root: %v", err)
+		}
+		// Check root is not modified
```
Comment (RJPercival) [inline crypto/data_formats_test.go]: `SignLogRoot` takes `test.root` by value, so couldn't modify it even if it wanted to. Presumably the implementation of `SignLogRoot` changed since this test was written. Either way, I don't think it's necessary to test that `SignLogRoot` doesn't do something it can't possibly do.

File: crypto/signer.go [hunk 9]
```diff
@@ -64,54 +53,39 @@ func NewSigner(sigAlgo sigpb.DigitallySigned_SignatureAlgorithm, signer crypto.S
 }
 
 // Sign obtains a signature after first hashing the input data.
-func (s Signer) Sign(data []byte) (sigpb.DigitallySigned, error) {
+func (s Signer) Sign(data []byte) (*sigpb.DigitallySigned, error) {
```
Comment (RJPercival) [inline crypto/signer.go]: I don't think you need to copy `s` when invoking this func.

File: crypto/signer.go [hunk 10]
```diff
@@ -64,54 +53,39 @@ func NewSigner(sigAlgo sigpb.DigitallySigned_SignatureAlgorithm, signer crypto.S
...
 	h := s.hash.New()
 	h.Write(data)
 	digest := h.Sum(nil)
 
 	if len(digest) != s.hash.Size() {
-		return sigpb.DigitallySigned{}, fmt.Errorf("hasher returned unexpected digest length: %d, %d",
+		return nil, fmt.Errorf("hasher returned unexpected digest length: %d, %d",
```
Comment (RJPercival) [inline crypto/signer.go]: Use the "got %d, want %d" format of error message to make it clear which of the two values is which.

File: crypto/signer.go [hunk 11]
```diff
@@ -64,54 +53,39 @@ func NewSigner(sigAlgo sigpb.DigitallySigned_SignatureAlgorithm, signer crypto.S
...
+	j, err := json.Marshal(obj)
 	if err != nil {
-		glog.Warningf("Signer failed to sign root: %v", err)
-		return sigpb.DigitallySigned{}, err
+		return nil, err
 	}
-
-	return signature, nil
+	hash := objecthash.CommonJSONHash(string(j))
+	sig, err := s.Sign(hash[:])
```
Comment (RJPercival) [inline crypto/signer.go]: Could just do `return s.Sign(hash[:])`

File: crypto/verifier.go [hunk 12]
```diff
@@ -49,8 +61,23 @@ func PublicKeyFromPEM(pemEncodedKey string) (crypto.PublicKey, error) {
...
 
+// VerifyObject verifies the output of SignObject.
+func VerifyObject(pub crypto.PublicKey, obj interface{}, sig *sigpb.DigitallySigned) error {
+	j, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	hash := objecthash.CommonJSONHash(string(j))
+
+	if err := Verify(pub, hash[:], sig); err != nil {
```
Comment (RJPercival) [inline crypto/verifier.go]: Could just `return Verify(pub, hash[:], sig)`

File: crypto/verifier.go [hunk 13]
```diff
@@ -20,19 +20,31 @@ import (
...
+	pemData, err := ioutil.ReadFile(keyFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read: %s. %v", keyFile, err)
+	}
+	return PublicKeyFromPEM(pemData)
+}
+
 // PublicKeyFromPEM converts a PEM object into a crypto.PublicKey
-func PublicKeyFromPEM(pemEncodedKey string) (crypto.PublicKey, error) {
+func PublicKeyFromPEM(pemEncodedKey []byte) (crypto.PublicKey, error) {
```
Comment (RJPercival) [inline crypto/verifier.go]: Isn't this reversing a review comment I made on the last PR, about PEM parameters having a consistent type?

File: crypto/data_formats.go (hunk 4 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: I added the TODO and assigned it to @AlCutter.
It's my opinion that we should remove everything in this file and make signing a generic thing rather than creating new functions for every object we want to sign.  But that should be a separate PR.  Just separating it here so it's easier to remove.

File: crypto/key_manager.go (hunk 1 above)
Comment (gdbelvin) [inline crypto/key_manager.go]: done

File: crypto/data_formats.go (hunk 6 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: Ie. Not with a generic object hash. -- ie. using the special map that is defined above.

File: crypto/data_formats.go (hunk 4 above)
Comment (RJPercival) [inline crypto/data_formats.go]: I think the TODO should have your name in it then. The name in a TODO isn't who it's assigned to, but who can provide the most context when someone wants to fix it. It sounds like that's you.

File: crypto/data_formats.go (hunk 6 above)
Comment (RJPercival) [inline crypto/data_formats.go]: Rephrasing the comment to say that would be fine :)

File: crypto/data_formats.go (hunk 4 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go (hunk 6 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+func HashTrillianSignedLogRoot(root trillian.SignedLogRoot) []byte {
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+	rootMap := make(map[string]interface{})
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go (hunk 5 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	rootMap[mapKeyTimestampNanos] = strconv.FormatInt(root.TimestampNanos, 10)
+	rootMap[mapKeyTreeSize] = strconv.FormatInt(root.TreeSize, 10)
+
+	hash := objecthash.ObjectHash(rootMap)
+	return hash[:]
+}
+
+// SignLogRoot updates a log root to include a signature from the crypto signer this object
+// was created with. Signatures use objecthash on a fixed JSON format of the root.
+func (s Signer) SignLogRoot(root trillian.SignedLogRoot) (*sigpb.DigitallySigned, error) {
```
Comment (gdbelvin) [inline crypto/data_formats.go]: done

File: crypto/data_formats_test.go (hunk 8 above)
Comment (gdbelvin) [inline crypto/data_formats_test.go]: done

File: crypto/signer.go (hunk 9 above)
Comment (gdbelvin) [inline crypto/signer.go]: done

File: crypto/signer.go (hunk 10 above)
Comment (gdbelvin) [inline crypto/signer.go]: done.  Removing this unneeded test.

File: crypto/signer.go (hunk 11 above)
Comment (gdbelvin) [inline crypto/signer.go]: done

File: crypto/verifier.go (hunk 13 above)
Comment (gdbelvin) [inline crypto/verifier.go]: yep.  Fixing now.

File: crypto/verifier.go (hunk 12 above)
Comment (gdbelvin) [inline crypto/verifier.go]: done

File: crypto/data_formats_test.go (hunk 7 above)
Comment (gdbelvin) [inline crypto/data_formats_test.go]: done

**Comment by gdbelvin** [comment]
PTAL

File: crypto/data_formats.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+		mapKeyRootHash:       base64.StdEncoding.EncodeToString(root.RootHash),
+		mapKeyTimestampNanos: strconv.FormatInt(root.TimestampNanos, 10),
+		mapKeyTreeSize:       strconv.FormatInt(root.TreeSize, 10)}
+
+	hash := objecthash.ObjectHash(rootMap)
+	return hash[:]
+}
+
+// SignLogRoot returns a signature using objecthash on a fixed JSON format of the root.
+func (s *Signer) SignLogRoot(root trillian.SignedLogRoot) (*sigpb.DigitallySigned, error) {
```
Comment (codingllama) [inline crypto/data_formats.go]: nit: Move this method to signer.go? I find it unusual for the method to be defined in a separate file from the type.

File: crypto/data_formats_test.go [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+package crypto
+
+import (
+	"testing"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/testonly"
+)
+
+func TestSignLogRoot(t *testing.T) {
```
Comment (codingllama) [inline crypto/data_formats_test.go]: Move test to signer_test.go?

File: crypto/data_formats_test.go [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+package crypto
+
+import (
+	"testing"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/testonly"
+)
+
```
Comment (codingllama) [inline crypto/data_formats_test.go]: Add a test for HashLogRoot?

It may not be useful to check specific hash values, but we definitely want to make sure that RootHash, TimestampNanos and TreeSize are part of the computation.

File: crypto/verifier.go [hunk 20]
```diff
@@ -20,17 +20,29 @@ import (
...
 	"math/big"
 
+	"github.com/benlaurie/objecthash/go/objecthash"
 	"github.com/google/trillian/crypto/sigpb"
 )
 
 // ErrVerify occurs whenever signature verification fails.
 var ErrVerify = errors.New("signature verification failed")
 
+// PublicKeyFromFile returns the public key contained in the keyFile.
```
Comment (codingllama) [inline crypto/verifier.go]: nit: Add ", in PEM format" ?

File: crypto/verifier.go [hunk 21]
```diff
@@ -49,8 +61,19 @@ func PublicKeyFromPEM(pemEncodedKey string) (crypto.PublicKey, error) {
 	return parsedKey, nil
 }
 
+// VerifyObject verifies the output of SignObject.
```
Comment (codingllama) [inline crypto/verifier.go]: nit: "verifies the output of SignObject". I'm not sure what that means. Is SignObject == obj? Please rephrase. We should probably mention that obj is marshaled to JSON and hashed with CommonJSONHash before verification.

File: crypto/data_formats_test.go (hunk 7 above)
Comment (codingllama) [inline crypto/data_formats_test.go]: I don't think this got addressed.

File: client/client.go [hunk 22]
```diff
@@ -115,12 +115,18 @@ func (c *LogClient) UpdateSTR(ctx context.Context) error {
...
 	}
 	resp, err := c.client.GetLatestSignedLogRoot(ctx, req)
-	str := resp.SignedLogRoot
 	if err != nil {
 		return err
 	}
-	// TODO(gdbelvin): Verify SignedLogRoot Signature
+	str := resp.SignedLogRoot
 
+	// Verify SignedLogRoot signature.
```
Comment (codingllama) [inline client/client.go]: nit: Needless comment?

File: crypto/data_formats.go [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+// Constants used as map keys when building input for ObjectHash. They must not be changed
+// as this will change the output of hashRoot()
+const (
+	mapKeyRootHash       string = "RootHash"
+	mapKeyTimestampNanos string = "TimestampNanos"
+	mapKeyTreeSize       string = "TreeSize"
+)
+
+// HashLogRoot hashes SignedLogRoot objects in a custom way
+// using the map of properties listed above.
```
Comment (codingllama) [inline crypto/data_formats.go]: nit: Please document the contract of HashLogRoot more explicit. The properties above are not public, therefore I don't think they should be referenced by documentation.

File: crypto/data_formats.go [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+	rootMap := map[string]string{
+		mapKeyRootHash:       base64.StdEncoding.EncodeToString(root.RootHash),
+		mapKeyTimestampNanos: strconv.FormatInt(root.TimestampNanos, 10),
+		mapKeyTreeSize:       strconv.FormatInt(root.TreeSize, 10)}
+
+	hash := objecthash.ObjectHash(rootMap)
+	return hash[:]
+}
+
+// SignLogRoot returns a signature using objecthash on a fixed JSON format of the root.
```
Comment (codingllama) [inline crypto/data_formats.go]: Suggestion: "SignLogRoot returns a signature over the HashLogRoot() of root"

File: crypto/verifier.go (hunk 21 above)
Comment (codingllama) [inline crypto/verifier.go]: Suggestion: "(...) verifies the output of Signer.SignObject"

File: crypto/data_formats.go (hunk 17 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: I'll remove this function.

File: crypto/data_formats_test.go (hunk 18 above)
Comment (gdbelvin) [inline crypto/data_formats_test.go]: done

File: crypto/data_formats_test.go (hunk 19 above)
Comment (gdbelvin) [inline crypto/data_formats_test.go]: done

File: crypto/verifier.go (hunk 20 above)
Comment (gdbelvin) [inline crypto/verifier.go]: done

File: client/client.go (hunk 22 above)
Comment (gdbelvin) [inline client/client.go]: na, I think it adds some structure to the verifications this is doing.

File: crypto/data_formats.go (hunk 24 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: removed this func

File: crypto/data_formats.go (hunk 23 above)
Comment (gdbelvin) [inline crypto/data_formats.go]: how's this:

```
// HashLogRoot hashes SignedLogRoot objects using ObjectHash with
// "RootHash", "TimestampNanos", and "TreeSize", used as keys in
// a map.
```

File: crypto/verifier.go (hunk 21 above)
Comment (gdbelvin) [inline crypto/verifier.go]: done

File: crypto/signer_test.go [hunk 25]
```diff
@@ -111,11 +111,45 @@ func TestSignLogRootSignerFails(t *testing.T) {
...
+	_, err := logSigner.Sign(HashLogRoot(root))
 
 	testonly.EnsureErrorContains(t, err, "signfail")
 }
 
 func createTestSigner(mock *MockSigner) *Signer {
 	return NewSigner(sigpb.DigitallySigned_RSA, mock)
 }
+
+func TestSignLogRoot(t *testing.T) {
```
Comment (codingllama) [inline crypto/signer_test.go]: Remove the test too? The method doesn't exist anymore.

File: testonly/integration/clientserver.go [hunk 26]
```diff
@@ -40,12 +40,15 @@ const (
 )
 
 var (
+	publicKeyFile      = "../testdata/log-rpc-server.pubkey.pem"
```
Comment (codingllama) [inline testonly/integration/clientserver.go]: I think @RJPercival added a comment here about using the constants of testonly/keys.go instead of reading files, but I can't find it anymore.  :/

Could you change it, please?

File: crypto/signer_test.go (hunk 25 above)
Comment (gdbelvin) [inline crypto/signer_test.go]: done

File: testonly/integration/clientserver.go (hunk 26 above)
Comment (gdbelvin) [inline testonly/integration/clientserver.go]: done

**Comment by gdbelvin** [comment]
PTAL

File: crypto/signer_test.go [hunk 27]
```diff
@@ -51,30 +50,29 @@ func (i usesSHA256Hasher) String() string {
 }
 
 func TestSigner(t *testing.T) {
```
Comment (codingllama) [inline crypto/signer_test.go]: Would you mind keeping the old test, in addition to the one you added? I think both add value in different areas.

Comment (gdbelvin) [inline crypto/signer_test.go]: I'll add some of the old test back in.

File: crypto/signer_test.go [hunk 28]
```diff
@@ -65,12 +66,18 @@ func TestSigner(t *testing.T) {
...
 	} {
-		signature, err := signer.Sign(test.message)
+		sig, err := signer.Sign(test.message)
+		if got, want := sig.HashAlgorithm, sigpb.DigitallySigned_SHA256; got != want {
+			t.Fatalf("Hash alg incorrect, got %v expected %d", got, want)
+		}
+		if got, want := sig.SignatureAlgorithm, sigpb.DigitallySigned_ECDSA; got != want {
+			t.Fatalf("Sig alg incorrect, got %v expected %v", got, want)
+		}
 		if err != nil {
```
Comment (codingllama) [inline crypto/signer_test.go]: Move `if err != nil` to line 70, add a `continue` after the `t.Errorf("Failed to sign ...")` below. We shouldn't assume `sig != nil` if `err != nil`.

Comment (codingllama) [inline crypto/signer_test.go]: Could you check that `len(sig.Signature) > 0`? With that I think I'm happy for you to drop the other test.

Comment (gdbelvin) [inline crypto/signer_test.go]: done

File: crypto/signer_test.go [hunk 29]
```diff
@@ -67,14 +67,17 @@ func TestSigner(t *testing.T) {
 		{message: []byte("message")},
 	} {
 		sig, err := signer.Sign(test.message)
+		if err != nil {
+			t.Errorf("Failed to sign log root: %v", err)
```
Comment (codingllama) [inline crypto/signer_test.go]: nit: Add a continue here? If signing failed there isn't much point in progressing the test.
